### ISSUE TITLE
packages.yaml versions only used if they exist.

### ISSUE THREAD

**Issue by citibeth** [spack/spack#2302]
@tgamblin 

Suppose I'm going to build (or have built) a package manually using the `spack setup` stuff.  Typically when I do that, I invent a fake version, I like to use `@local`.  (Unlike `@develop`, Spack does absolutely nothing special with `@local`).  These fake versions are not being picked up from `packages.yaml`.  Witness:

Good:
```
$ spack spec libxml2
libxml2@2.9.4%gcc@4.9.3~python arch=linux-centos7-x86_64
    ^xz@5.2.2%gcc@4.9.3 arch=linux-centos7-x86_64
    ^zlib@1.2.8%gcc@4.9.3 arch=linux-centos7-x86_64
```

Now I want to setup libxml2 for manually compile... Good:
```
$ spack spec libxml2@local
libxml2@local%gcc@4.9.3~python arch=linux-centos7-x86_64
    ^xz@5.2.2%gcc@4.9.3 arch=linux-centos7-x86_64
    ^zlib@1.2.8%gcc@4.9.3 arch=linux-centos7-x86_64
```

Suppose I want to set up libxml2 and xz for manual compile:  Good:
```
$ spack spec libxml2@local ^xz@local
libxml2@local%gcc@4.9.3~python arch=linux-centos7-x86_64
    ^xz@local%gcc@4.9.3 arch=linux-centos7-x86_64
    ^zlib@1.2.8%gcc@4.9.3 arch=linux-centos7-x86_64
```

But... suppose I put the following in my `packages.yaml`:
```
packages:
    xz:
        version: [local]
```

This is not picked up, and I get the same as if it were not there:
```
$ spack spec libxml2@local
libxml2@local%gcc@4.9.3~python arch=linux-centos7-x86_64
    ^xz@5.2.2%gcc@4.9.3 arch=linux-centos7-x86_64
    ^zlib@1.2.8%gcc@4.9.3 arch=linux-centos7-x86_64
```

**Comment by citibeth** [comment]
Admittedly... I'm not sure if this is a bug or a feature.  Fixing it probably requires more thinking than just making it do what I first expected in this case.

**Comment by alalazo** [comment]
Closing, since this seems to work as designed. Most preferences in `packages.yaml` are soft, in the sense that they are respected if possible but will not generate an error otherwise. If this issue was asking for having installed specs considered in the concretization process, the same feature is tracked also in multiple other places (a solution will be implemented after #19501 is merged)
